File: boj/python/string/boj17609.py
import sys


def solution():
    result = 2
    length = len(s)

    start = 0
    end = length - 1

    if check_palindrome(s, start, end):
        return 0

    while start < end:
        if s[start] == s[end]:
            start += 1
            end -= 1
        else:
            left = check_palindrome(s, start + 1, end)
            right = check_palindrome(s, start, end - 1)
            if left or right:
                result = 1
            break
    return result


def check_palindrome(string, left, right):
    while left < right:
        if string[left] != string[right]:
            return False
        left += 1
        right -= 1
    return True


if __name__ == "__main__":
    input = sys.stdin.readline
    for _ in range(int(input())):
        s = input().rstrip()
        print(solution())

File: boj/java/src/string/Boj17609.java
package string;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Boj17609 {
	int solve(String input) {
		int result = 2;
		int start = 0;
		int end = input.length() - 1;

		if (checkPalindrome(input, start, end)) {
			return 0;
		}

		while (start < end) {
			if (input.charAt(start) == input.charAt(end)) {
				start++;
				end--;
			} else {
				boolean left = checkPalindrome(input, start + 1, end);
				boolean right = checkPalindrome(input, start, end - 1);
				if (left || right) {
					result = 1;
				}
				break;
			}
		}

		return result;
	}

	boolean checkPalindrome(String input, int left, int right) {
		while (left < right) {
			if (input.charAt(left++) != input.charAt(right--)) {
				return false;
			}
		}
		return true;
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		Boj17609 solution = new Boj17609();

		for (int i = 0; i < n; i++) {
			String s = br.readLine();
			System.out.println(solution.solve(s));
		}
	}
}
